src/kagglehub/models_helpers.py

```diff
@@ -0,0 +1,227 @@
+import datetime
+import logging
+import os
+import shutil
+import tempfile
+from http import HTTPStatus
+from typing import Optional
+
+import requests
+
+from kagglehub.clients import KaggleApiV1Client
+from kagglehub.exceptions import BackendError, KaggleApiHTTPError
+from kagglehub.handle import ModelHandle
+
+logger = logging.getLogger(__name__)
+
+DATASET_METADATA_FILE = "dataset-metadata.json"
+OLD_DATASET_METADATA_FILE = "datapackage.json"
+KERNEL_METADATA_FILE = "kernel-metadata.json"
+MODEL_METADATA_FILE = "model-metadata.json"
+MODEL_INSTANCE_METADATA_FILE = "model-instance-metadata.json"
+
+
+def parse(string):
+    time_formats = ["%Y-%m-%dT%H:%M:%S", "%Y-%m-%dT%H:%M:%SZ", "%Y-%m-%dT%H:%M:%S.%f", "%Y-%m-%dT%H:%M:%S.%fZ"]
+    for t in time_formats:
+        try:
+            result = datetime.strptime(string[:26], t).replace(microsecond=0)
+            return result
+        except:  # noqa: E722, S110
+            pass
+    return string
+
+
+class DirectoryArchive:
+    def __init__(self, fullpath, format):  # noqa: A002
+        self._fullpath = fullpath
+        self._format = format
+        self.name = None
+        self.path = None
+
+    def __enter__(self):
+        self._temp_dir = tempfile.mkdtemp()
+        _, dir_name = os.path.split(self._fullpath)
+        self.path = shutil.make_archive(os.path.join(self._temp_dir, dir_name), self._format, self._fullpath)
+        _, self.name = os.path.split(self.path)
+        return self
+
+    def __exit__(self, *args):
+        shutil.rmtree(self._temp_dir)
+
+
+class File(object):  # noqa: UP004
+    def __init__(self, init_dict):
+        parsed_dict = {k: parse(v) for k, v in init_dict.items()}
+        self.__dict__.update(parsed_dict)
+        self.size = File.get_size(self.totalBytes)
+
+    def __repr__(self):
+        return self.ref
+
+    @staticmethod
+    def get_size(size, precision=0):
+        suffixes = ["B", "KB", "MB", "GB", "TB"]
+        suffix_index = 0
+        while size >= 1024 and suffix_index < 4:  # noqa: PLR2004
+            suffix_index += 1
+            size /= 1024.0
+        return "%.*f%s" % (precision, size, suffixes[suffix_index])
+
+
+def _upload_blob(file_path):
+    data = {
+        "type": "model",
+        "name": os.path.basename(file_path),
+        "contentLength": os.path.getsize(file_path),
+        "lastModifiedEpochSeconds": int(os.path.getmtime(file_path)),
+    }
+    api_client = KaggleApiV1Client()
+    response = api_client.post("/blobs/upload", data=data)
+    if "error" in response and response["error"] != "":
+        raise BackendError(response["error"])
+
+    with open(file_path, "rb") as f:
+        file_data = f.read()
```

Comment:
Many files that you will upload will be too large to fit in memory. This will OOM for any file > than your available RAM. You should stream it in chunk to the put request instead.